Fundamentos de Go

Golang (o Go) es un lenguaje de código abierto de Google, diseñado para la simplicidad, alto rendimiento y concurrencia. Es compilado, de tipado estático y eficiente en memoria. Sus fundamentos clave incluyen una sintaxis limpia tipo C, manejo de paquetes, estructuras en lugar de clases, y rutinas (goroutines) para concurrencia. 

1. Estructura de un archivo en Go

Todo archivo de Go debe pertenecer a un package. El punto de entrada de cualquier aplicación ejecutable es el paquete main y la función main(). La sentencia import permite traer funcionalidades de la Standard Library o paquetes externos.

In [1]:
package main // Define el nombre del paquete actual

import "fmt" // Importa el paquete 'fmt' para entrada/salida formateada

func main() {
    // El punto de entrada del programa
    fmt.Println("¡Hola! Bienvenido a Go.")
}

¡Hola! Bienvenido a Go.


2. Declaración de Variables y Constantes

Go es un lenguaje de tipado estático. Ofrece dos formas principales de declarar variables: el uso de la palabra clave var (útil para declaraciones explícitas o a nivel de paquete) y el operador de inferencia corta := (solo dentro de funciones).

    • var: Su uso principal es Global o local, permite declarar sin inicializar. Su inferencia de tipo es Opcional.

    • := : Su uso principal es solo local (dentro de funciones), requiere inicialización. Su inferencia de tipo es Automatica.

    • cosnt: Se asiga valores que no cambian durante la ejecucion. Su inferencia es Estricta.

Nota: En Go, si declaras una variable y no la usas, el compilador arrojará un error. Esto mantiene el código limpio y eficiente.


In [2]:
package main

import "fmt"

const Pi = 3.1415 // Constante: no puede ser reasignada

func main() {
    var nombre string = "Gopher" // Declaración explícita
    edad := 15                   // Declaración corta con inferencia de tipo (int)
    
    fmt.Printf("Nombre: %s, Edad: %d, Pi: %f\n", nombre, edad, Pi)
}

Nombre: Gopher, Edad: 15, Pi: 3.141500


3. Tipos de Datos Básicos

Go prioriza la seguridad y el rendimiento. Los tipos de datos se dividen principalmente en números, cadenas de texto y booleanos.

• int: Entero (depende de la arquitectura: 32 o 64 bits). (42)
• float64: Número de punto flotante de 64 bits. (3.14)
• string: Cadena de caracteres (UTF-8). ("Hola")
• bool: Valor lógico. (true / false)

In [ ]:
package main

import "fmt"

func main() {
    var activo bool = true
    var precio float64 = 199.99
    var simbolo rune = 'G' // 'rune' es un alias para int32, representa un punto de código Unicode
    
    fmt.Printf("¿Activo?: %t, Precio: %.2f, Run: %c\n", activo, precio, simbolo)
}

4. Estructuras de Control

Go simplifica la lógica eliminando los paréntesis en las condiciones y reduciendo los bucles a una única palabra clave: for.

Nota: La llave de apertura { debe estar obligatoriamente en la misma línea que la sentencia de control (if, for, switch). Si se coloca en la línea siguiente, el compilador insertará un punto y coma erróneo.

In [ ]:
package main

import "fmt"

func main() {
    // if / else
    puntos := 85
    if puntos >= 90 {
        fmt.Println("Excelente")
    } else {
        fmt.Println("Sigue intentando")
    }

    // El único bucle: for
    fmt.Print("Bucle for: ")
    for i := 0; i < 3; i++ {
        fmt.Printf("%d ", i)
    }

    // switch (no requiere 'break' explícito)
    dia := "Lunes"
    switch dia {
    case "Lunes":
        fmt.Println("\nInicio de semana")
    default:
        fmt.Println("\nOtro día")
    }
}

5. Colecciones: Slices y Maps

A diferencia de los arrays (tamaño fijo), los Slices son estructuras dinámicas y mucho más comunes en Go. Los Maps son estructuras de clave-valor (diccionarios).

In [ ]:
package main

import "fmt"

func main() {
    // Slices: Dinámicos y potentes
    frutas := []string{"Manzana", "Banana"}
    frutas = append(frutas, "Uva") // Añadir elementos
    
    // Maps: Clave-Valor
    capitales := map[string]string{
        "Francia": "París",
        "Italia":  "Roma",
    }

    fmt.Println("Slice de frutas:", frutas)
    fmt.Println("Capital de Italia:", capitales["Italia"])
}

6. Funciones y Retornos Múltiples

Las funciones en Go son ciudadanos de primera clase. Una de sus características más potentes es la capacidad de retornar múltiples valores, lo que se usa convencionalmente para el manejo de errores.

In [ ]:
package main

import (
    "fmt"
    "errors"
)

// Función con parámetros y múltiples retornos (valor, error)
func dividir(a, b float64) (float64, error) {
    if b == 0 {
        return 0, errors.New("no se puede dividir por cero")
    }
    return a / b, nil // 'nil' representa la ausencia de valor (null)
}

func main() {
    resultado, err := dividir(10, 2)
    
    if err != nil {
        fmt.Println("Error:", err)
    } else {
        fmt.Printf("Resultado: %.2f\n", resultado)
    }
}

7. Structs e Interfaces: El Modelo de Datos en Go

Go no tiene clases ni herencia. En su lugar, utiliza Structs (estructuras) para agrupar datos y Métodos para definir comportamiento. Las Interfaces son contratos: si un tipo implementa los métodos de una interfaz, entonces "es" de ese tipo (Duck Typing implícito).

    • Struct: Define la forma de los datos (campos).
    • Método: Función asociada a un struct mediante un receiver.
    • Interfaz: Define un conjunto de firmas de métodos (comportamiento).

Nota: En Go, el uso de punteros (*) en los métodos es fundamental. Si quieres que un método modifique los datos del struct original, debes usar un puntero como receptor.


In [ ]:
package main

import "fmt"

// Definición de una interfaz (contrato)
type Saludador interface {
    Saludar() string
}

// Definición de un struct
type Persona struct {
    Nombre string
    Edad   int
}

// Método asociado al struct Persona
// Implementa implícitamente la interfaz Saludador
func (p Persona) Saludar() string {
    return fmt.Sprintf("Hola, soy %s y tengo %d años.", p.Nombre, p.Edad)
}

func main() {
    p := Persona{Nombre: "Elena", Edad: 30}
    
    // Podemos tratar a 'p' como un 'Saludador'
    var s Saludador = p
    fmt.Println(s.Saludar())
}

8. Concurrencia: Goroutines y Channels

La concurrencia es nativa en Go. Una Goroutine es un hilo ligero gestionado por el runtime de Go. Los Channels (canales) son los conductos que permiten a las goroutines comunicarse y sincronizarse de forma segura.

Nota: "No te comuniques compartiendo memoria; comparte memoria comunicándote". Esta es la máxima de Go para evitar condiciones de carrera (race conditions).

In [ ]:
package main

import (
    "fmt"
    "time"
)

// Función que se ejecutará de forma concurrente
func enviarMensaje(canal chan string) {
    time.Sleep(2 * time.Second) // Simulamos una tarea pesada
    canal <- "Mensaje desde la Goroutine" // Enviamos datos al canal
}

func main() {
    // Creamos un canal de tipo string
    mensajeCanal := make(chan string)

    // Iniciamos la goroutine con la palabra clave 'go'
    go enviarMensaje(mensajeCanal)

    fmt.Println("Esperando el mensaje...")

    // La ejecución se bloquea aquí hasta recibir el dato
    mensaje := <-mensajeCanal 
    fmt.Println("Recibido:", mensaje)
}

9. Manejo de Errores y Defer

Go no utiliza excepciones (try-catch). En su lugar, el manejo de errores es explícito. Además, cuenta con la palabra clave defer, que asegura que una función se ejecute al final del ámbito actual (ideal para cerrar archivos o conexiones).

In [ ]:
package main

import "fmt"

func procesoCritico() {
    // defer se ejecutará justo antes de salir de la función, pase lo que pase
    defer fmt.Println("Limpieza de recursos completada.")

    fmt.Println("Ejecutando lógica principal...")
    
    // Aquí podría haber un error, pero el defer garantiza la limpieza
}

func main() {
    procesoCritico()
}